pkgs/top-level/release-outpaths-parallel.nix  5x faster eval checks

amjoseph-nixpkgs opened the pull request:
This PR is implemented entirely in pure Nix code and a single call to `xargs`.

- It has already been used to [build the entire release packageset](https://github.com/NixOS/nixpkgs/pull/268097) at larger-than-Hydra scale.
  - This is a complete, working solution, not a component or a theoretical possibility.
- It works with any Nix interpreter, and is not cppnix-specific like Hydra is.
  - It does not need to link against any cppnix shared libraries.
  - It does not use cppnix's internal interfaces.
- It is in-tree as part of nixpkgs
  - It does not add additional out-of-repo dependencies with their own gatekeeping committees.



## Results

On a 32-core machine, the following command takes 4m58s:

```
NIX_BUILD_CORES=32 time pkgs/top-level/release-outpaths-parallel.sh false
```

and produces **exactly** the same output as this command (the one ofborg uses), which takes 24m10s:

```
time nix-env -qaP --no-name --out-path --arg checkMeta false -f pkgs/top-level/release-outpaths.nix | sort
```

... in 1/5th of the wall-clock time.  So the eval-check latency is reduced by 5x.  With `checkMeta=true` the ratio is almost identical: 37m54s with the current method, 7m38s with `release-outpaths-parallel`.

There is room to improve even further:

- Half of the 4m58s is spent in the (completely serial) `release-attrpaths-superset.nix` process.  However that process produces output incrementally, meaning that the parallel part of the eval-checks can be started as soon as the first line of `release-attrpaths-superset.nix` output is produced -- this should give another 25% improvement.
- Right now the `4*NUM_CORES` chunks are scheduled blindly, by simply taking alphabetical chunks of job attrpaths.  This results in a huge amount of duplicated work.  Simple dependency-based scheduling (possibly using the previous commit's attrpath dependency graph as an approximation of the current one) should give another large boost, possibly even 50%.

Includes

- https://github.com/NixOS/nixpkgs/pull/269356

Commit:
pkgs/top-level/release-outpaths.nix: vendor from ofborg

This commit vendors `outpaths.nix` from ofborg commit

  74f38efa7ef6f0e8e71ec3bfc675ae4fb57d7491

Commit:
test.cuda: fix broken eval

Commit:
treewide: add __attrsFailEvaluation and __recurseIntoDerivationForReleaseJobs

Commit:
pkgs/top-level/release-outpaths.nix: add attrNamesOnly option

Commit:
pkgs/top-level/release.nix: add attrNamesOnly option

Commit:
pkgs/top-level/all-packages.nix: remove self-referential top-level attributes

Commit:
pkgs/top-level/release-attrpaths-superset.nix: init

This file walks the entire nixpkgs tree and emits a superset of all
release attrnames in only 44 seconds on a 3ghz CPU, using 5 gbytes
of memory.  By comparison, on the same CPU the `nix-env` hack used
by ofborg on every PR submission requires 41 *minutes* and peaks at
60 gbytes, even with checkMeta turned off.  Full details below.

This is:

- 46x faster (or 2.1% of the elapsed time)
- 12.5x less memory (or 8.0% of the peak memory usage)

In order to replace the ofborg check, this list of attrnames must
then be post-filtered for platform-relevance.  However, crucially,
the post-filtering can be done *in parallel* on multiple cores by
splitting the attrname list in to chunks.  Generating the list of
attrnames cannot be parallelized because it is a single-threaded
cppnix task.

This PR also adds `recurseForDerivations` where necessary within
nixpkgs in order to make this possible -- it screens out various
non-tryEval-catchable failures and infinite recursions.  Before
undraftifying, I will add an invocation of this command to the CI
tests, to ensure that the work performed here is not immediately
undone.  My next PR will then add an additional CI check confirming
that the emitted attrpaths are in fact a superset of the release
attrpaths calculated by the slow-memory-hog ofborg method.

I have manually confirmed that this is the case at the tip commit of
this PR, but we need CI to make sure this remains true until ofborg
switches to this more-efficient method of calculation; at that point
the superset-check can be dropped.

According to GNU Time,

        Command being timed: "nix-instantiate --eval --strict --json pkgs/top-level/release-attrpaths-superset.nix -A names"
        User time (seconds): 44.88
        System time (seconds): 8.09
        Percent of CPU this job got: 99%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:53.20
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 4823028
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 3611240
        Voluntary context switches: 113
        Involuntary context switches: 949
        Swaps: 0
        File system inputs: 1480
        File system outputs: 5944
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0

Compared to release-outpaths.nix:

        Command being timed: "nix-env -qaP --no-name --out-path --arg checkMeta false --argstr path /git/work/pr/release-outpaths -f pkgs/top-level/release-outpaths.nix"
        User time (seconds): 2120.67
        System time (seconds): 337.80
        Percent of CPU this job got: 98%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 41:37.91
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 60171768
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 2
        Minor (reclaiming a frame) page faults: 230608113
        Voluntary context switches: 8876
        Involuntary context switches: 22275
        Swaps: 0
        File system inputs: 62624
        File system outputs: 72
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0

Commit:
AAAAAASomeThingsFailToEvaluate: provide a message which is actually helpful

Commit:
pkgs/top-level/release-outpaths.nix: adjust default path value

Commit:
pkgs/top-level/release-outpaths-parallel.{nix,sh}: init 5x faster eval checks

On a 32-core machine, the following command takes 4m58s:

  NIX_BUILD_CORES=32 time pkgs/top-level/release-outpaths-parallel.sh false

Produces exactly the same output as this command, which takes 24m10s:

  time nix-env -qaP --no-name --out-path --arg checkMeta false -f pkgs/top-level/release-outpaths.nix | sort

... in 1/5th of the wall-clock time.  So the eval-check latency is
reduced by 5x.

Includes

- https://github.com/NixOS/nixpkgs/pull/269356

pkgs/top-level/release-outpaths-parallel.nix: pass includeBroken

Commit:
lib/tests/release.nix: run release-attrpaths-superset.nix

This verifies that release-attrpaths-superset.nix does not encounter
infinite recursion or non-tryEval-able failures.

Commit:
pkgs/test/release/release-attrpaths-superset.nix: init

This test checks that what release-attrpaths-superset.nix calculates
is in fact a superset of the attributes which are calculated by the
much-slower ofborg method.

To run the test, invoke this command:

  nix-build pkgs/test/release/release-attrpaths-superset.nix -A failures

The result should be the empty list [].

Artturin commented on pkgs/top-level/release-outpaths-parallel.sh:
```diff
@@ -0,0 +1,13 @@
+
+NUM_CHUNKS=$((4*$NIX_BUILD_CORES))
+
+parallel -j $NIX_BUILD_CORES \
```
Did you try nix-eval-jobs, it does parallel but doesn't help with memory usage AFAIK

Mic92 replied:
It helps with memory usage as well because it restarts workers if they take too much memory. The chunking at the moment might be less ideal because it doesn't chunk the list into X parts and gives this to threads but instead starts to iterate the list from the beginning and give each thread an attribute in lexical key order. I think it would be faster if attributes closer to each other would be evaluated by the same thread to avoid evaluating the same thing twice (i.e. x86_64-linux packages evaluated on thread1 and aarch64-linux packages on thread2).

Mic92 replied:
However from my tests I would say that having more threads than CPUs will actually slow you down because you end up evaluating the same thing more. I usually have less threads than cores for large package sets like nixpkgs to avoid duplicating work and rather trait it for memory.

Commit:
pkgs/top-level/release-outpaths.nix: simplify invocation

Commit:
pkgs/top-level/release-outpaths.nix: add includeBroken parameter

Commit:
pkgs/top-level/release-outpaths.nix: make systems parameter optional

Commit:
pkgs/top-level/nohydra: init

cole-h commented:
(Sorry, GitHub mobile fat finger...... Love how there's no confirmation...)

adisbladis commented:
I would like to note that using this in ofborg would mean giving up the performance report, at least as it is today.

adisbladis commented:
Conceptually this PR works similarly to how [nix-eval-jobs](https://github.com/nix-community/nix-eval-jobs) works, just implemented as a pile of shell/nix instead of using the evaluator directly.
I think it's better to make ofborg use nix-eval-jobs.

cc https://github.com/NixOS/ofborg/issues/440

Commit:
pkgs/top-level/release-attrpaths-superset.nix: add enableWarnings?true parameter

Commit:
pkgs/top-level/release-attrpaths-superset.nix: exempt AAAAAASomeThingsFailToEvaluate from warnings

Commit:
pkgs/top-level/release-attrpaths-superset.nix: pass through checkMeta

Commit:
pkgs/top-level/release-outpaths.nix: dont short-circuit handleEvalIssue when attrNamesOnly

Commit:
pkgs/top-level/release-outpaths.nix: never attempt to build unfree packages, most of them are broken

Since Hydra does not build unfree packages an astonishing proportion
of them are broken yet not marked meta.broken.

Commit:
pkgs/top-level/release-outpaths-parallel.nix: remove default for attrPathFile argument

Commit:
pkgs/top-level/release-attrpaths-superset.nix: add another weird edge case

Ugh, not only do we have jobsets-within-derivations, we have
situations where a meta.broken==true derivation contains
meta.broken==false jobsets which are in fact broken but nobody
noticed because they are meta.license==unfree.

Adding an edge case for now; we should mark these packages broken
and remove the edge case.

One example is promscale_extension.

Commit:
pkgs/top-level/nohydra/instantiate.sh: dont attempt to build unfree software, most of it is broken

Mic92 commented on pkgs/top-level/release-outpaths-parallel.nix:
```diff
@@ -0,0 +1,56 @@
+Invocation; note that the number of processes spawned is four times
+the number of cores -- this helps in two ways:
+
+1. Keeping cores busy while I/O operations are in flight
```
Interesting so far I have not seen nix evaluation to be significantly I/O bound.

amjoseph-nixpkgs replied:
This may be specific to my setup.

The machine from which I launch builds has *two* Nix stores: a fast 3TB nvme `/nix/store` and a huge raid-0 spinning-rust secondary store (which is a `trusted-substituter`).  I never run GC on the spinning-rust store.

I want to switch to a single store using bcachefs, but I'm still testing it out on less critical applications.

amjoseph-nixpkgs commented:
> Conceptually this PR works similarly to how [nix-eval-jobs](https://github.com/nix-community/nix-eval-jobs) works, just implemented as a pile of shell/nix instead of using the evaluator directly.

Conceptually nix-eval jobs works similarly to how this PR works, just implemented as a pile of C code instead of 99% pure Nix plus a one-line command `nix-instantiate | jq | xargs` pipeline (which isn't in Nix because `recursive-nix` isn't finished yet).

This PR has the following advantages:

- It works with any Nix interpreter, rather than being cppnix-specific.
- It has already been used to [build the entire release packageset](https://github.com/NixOS/nixpkgs/pull/268097) at larger-than-Hydra scale
- It is in-tree as part of nixpkgs

The last point is non-negotiable.  If you want your project to be used instead of this, you must first move it in-tree.  The goal is to have *fewer* out-of-monorepo CT pieces, not *more* of them.

Mic92 commented:
I don't agree that all tooling that we use in nixpkgs also has to be part of nixpkgs. This repository is already too big and we have issues tracking pull requests properly. Many of these tools also have their own, smaller CI setup that maybe even test multiple version of nixpkgs. Monorepos don't work well beyond a certain scale for opensource projects because opensource communities don't have the resources to re-invent all their tooling to scale to large repositories.

amjoseph-nixpkgs commented:
The tests need to be in the monorepo.

Commit:
delete obsolete leftover shell script

Commit:
nohydra: fix the sagetex problem

bew commented:
Why did you closed it?
Is it superseded by anything else?

katexochen commented:
@bew I think this PR was closed in in the light of https://github.com/NixOS/nixpkgs/issues/50105#issuecomment-1902682831. While not necessary, Adam chose to closed some of their open PRs (which is of cause okay if they like to do so).

Mic92 commented:
Should we revert it than until someone else picks it up?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-exclude-derivation-in-a-package-closure-from-the-binary-cache/50144/9

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/infrastructure-announcement-the-future-of-ofborg-your-help-needed/56025/37

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/infrastructure-announcement-the-future-of-ofborg-your-help-needed/56025/39
